src/cs/Ssh/SecureStream.cs

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DevTunnels.Ssh.Events;
+
+namespace Microsoft.DevTunnels.Ssh;
+
+/// <summary>
+/// Establishes an end-to-end encrypted two-way authenticated data stream over an underlying
+/// transport stream, using the SSH protocol but providing simplified interface that is limited to
+/// a single duplex stream (channel).
+/// </summary>
+/// <remarks>
+/// This class is a complement to <see cref="MultiChannelStream"/>, which provides only the
+/// channel-multiplexing functions of SSH.
+/// 
+/// To establish a secure connection, the two sides first establish an insecure transport stream
+/// over a pipe, socket, or anything else. Then they encrypt and authenticate the connection
+/// before beginning to send and receive data.
+/// </remarks>
+public class SecureStream : Stream
+{
+	private readonly Stream transportStream;
+	private readonly SshSession session;
+	private readonly SshClientCredentials? clientCredentials;
+	private readonly SshServerCredentials? serverCredentials;
+	private SshStream? stream;
+
+	/// <summary>
+	/// Creates a secure stream over an underlying transport stream, using client credentials
+	/// to authenticate.
+	/// </summary>
+	/// <param name="transportStream">Underlying (insecure) transport stream.</param>
+	/// <param name="clientCredentials">Client authentication credentials.</param>
+	/// <param name="trace">Optional trace source for SSH protocol tracing.</param>
+	public SecureStream(
+		Stream transportStream,
+		SshClientCredentials clientCredentials,
+		TraceSource? trace = null)
+	{
+		this.transportStream = transportStream ??
+			throw new ArgumentNullException(nameof(transportStream));
+		this.session = new SshClientSession(
+			SshSessionConfiguration.Default,
+			trace ?? new TraceSource(nameof(MultiChannelStream)));
+
+		this.session.Authenticating += OnSessionAuthenticating;
+		this.session.Closed += OnSessionClosed;
+
+		this.clientCredentials = clientCredentials ??
+			throw new ArgumentNullException(nameof(clientCredentials));
+	}
+
+	/// <summary>
+	/// Creates a secure stream over an underlying transport stream, using server credentials
+	/// to authenticate.
+	/// </summary>
+	/// <param name="transportStream">Underlying (insecure) transport stream.</param>
+	/// <param name="serverCredentials">Server authentication credentials.</param>
+	/// <param name="trace">Optional trace source for SSH protocol tracing.</param>
+	public SecureStream(
+		Stream transportStream,
+		SshServerCredentials serverCredentials,
+		TraceSource? trace = null)
+	{
+		this.transportStream = transportStream ??
+			throw new ArgumentNullException(nameof(transportStream));
+		this.session = new SshServerSession(
+			SshSessionConfiguration.Default,
```

Comment:
Is reconnection going to be a concern? Currently ssh session from the tunnel client to the tunnel host support reconnection. Will this be supported with the new model?

Reply:
I'm working on that, but it will be tricky when some channels are unencrypted and intercepted (potentially even modified?) by the relay, while other channels are E2E-encrypted separately. Therefore all the channels will need to be reconnected separately.

I'll probably add reconnect support to SecureStream later. We'll need a way to reconnect unencrypted channels also... the challenge is without the SSH packet sequencing and HMAC it's difficult to guarantee integrity of the data after reconnection. Maybe we'd use SecureStream with no encryption but with HMAC. (Then the relay could observe but not modify the traffic.)

But not for this PR.